### ISSUE TITLE
read_csv ValueError: Unknown protocol c

### ISSUE THREAD

**Issue by kumarivin** [dask/dask#1415]
I am getting an unknown protocol c error while trying to read_csv into the data frame.
My code is below
```
import dask.dataframe as dd
df = dd.read_csv('C:/Users/TG/Downloads/yoochoose-dataFull/yoochoose-clicks100.csv')
```
error is below

````
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-18-00d298f2858a> in <module>()
      1 import dask.dataframe as dd
----> 2 df = dd.read_csv('C:/Users/TG/Downloads/yoochoose-dataFull/yoochoose-clicks100.csv')
      3 get_ipython().magic(u'time len(df)')

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\dataframe\csv.pyc in read_csv(urlpath, blocksize, chunkbytes, collection, lineterminator, compression, sample, enforce, storage_options, **kwargs)
    220                                          sample=sample,
    221                                          compression=compression,
--> 222                                          **(storage_options or {}))
    223     if not isinstance(values[0], (tuple, list)):
    224         values = [values]

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\bytes\core.pyc in read_bytes(urlpath, delimiter, not_zero, blocksize, sample, compression, **kwargs)
     67                                             inherit_storage_options=kwargs)
     68     protocol = storage_options.pop('protocol')
---> 69     ensure_protocol(protocol)
     70     try:
     71         read_bytes = _read_bytes[protocol]

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\bytes\core.pyc in ensure_protocol(protocol)
    225 
    226     else:
--> 227         raise ValueError("Unknown protocol %s" % protocol)

ValueError: Unknown protocol c
````

**Comment by nirizr** [comment]
Totally a guess here but could you try using ´c:\´  and backslashes instead of forward slashes troughout the file path?

**Comment by mrocklin** [comment]
Can you verify that you are on dask version 0.10.1 ?

If not try 

    conda install -c conda-forge dask 

or 

    pip install dask[dataframe] --upgrade

**Comment by kumarivin** [comment]
upgraded as mentioned by @mrocklin  and changed path as mentioned by @nirizr .. I am getting another error as shown below. The file is a 1.5 gb file and sample data looks like this

sessioid,timestamp,itemid,category
1,2014-04-07T10:51:09.277Z,214536502,0
1,2014-04-07T10:54:09.868Z,214536500,0
1,2014-04-07T10:54:46.998Z,214536506,0
1,2014-04-07T10:57:00.306Z,214577561,0
2,2014-04-07T13:56:37.614Z,214662742,0
2,2014-04-07T13:57:19.373Z,214662742,0
2,2014-04-07T13:58:37.446Z,214825110,0
2,2014-04-07T13:59:50.710Z,214757390,0


```

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-26-b58b9e60e4fa> in <module>()
      2 import pandas as pd
      3 df = dd.read_csv('C:\Users\TG\Downloads\yoochoose-dataFull\yoochoose-clicks.dat')
----> 4 get_ipython().magic(u'time len(df)')

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\IPython\core\interactiveshell.pyc in magic(self, arg_s)
   2161         magic_name, _, magic_arg_s = arg_s.partition(' ')
   2162         magic_name = magic_name.lstrip(prefilter.ESC_MAGIC)
-> 2163         return self.run_line_magic(magic_name, magic_arg_s)
   2164 
   2165     #-------------------------------------------------------------------------

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\IPython\core\interactiveshell.pyc in run_line_magic(self, magic_name, line)
   2082                 kwargs['local_ns'] = sys._getframe(stack_depth).f_locals
   2083             with self.builtin_trap:
-> 2084                 result = fn(*args,**kwargs)
   2085             return result
   2086 

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\IPython\core\magics\execution.pyc in time(self, line, cell, local_ns)

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\IPython\core\magic.pyc in <lambda>(f, *a, **k)
    191     # but it's overkill for just that one bit of state.
    192     def magic_deco(arg):
--> 193         call = lambda f, *a, **k: f(*a, **k)
    194 
    195         if callable(arg):

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\IPython\core\magics\execution.pyc in time(self, line, cell, local_ns)
   1171         if mode=='eval':
   1172             st = clock2()
-> 1173             out = eval(code, glob, local_ns)
   1174             end = clock2()
   1175         else:

<timed eval> in <module>()

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\dataframe\core.pyc in __len__(self)
    286 
    287     def __len__(self):
--> 288         return reduction(self, len, np.sum, token='len').compute()
    289 
    290     def map_partitions(self, func, columns=no_default, *args, **kwargs):

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\base.pyc in compute(self, **kwargs)
     84             Extra keywords to forward to the scheduler ``get`` function.
     85         """
---> 86         return compute(self, **kwargs)[0]
     87 
     88     @classmethod

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\base.pyc in compute(*args, **kwargs)
    177         dsk = merge(var.dask for var in variables)
    178     keys = [var._keys() for var in variables]
--> 179     results = get(dsk, keys, **kwargs)
    180 
    181     results_iter = iter(results)

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\threaded.pyc in get(dsk, result, cache, num_workers, **kwargs)
     55     results = get_async(pool.apply_async, len(pool._pool), dsk, result,
     56                         cache=cache, queue=queue, get_id=_thread_get_id,
---> 57                         **kwargs)
     58 
     59     return results

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\async.pyc in get_async(apply_async, num_workers, dsk, result, cache, queue, get_id, raise_on_exception, rerun_exceptions_locally, callbacks, **kwargs)
    482                 _execute_task(task, data)  # Re-execute locally
    483             else:
--> 484                 raise(remote_exception(res, tb))
    485         state['cache'][key] = res
    486         finish_task(dsk, key, state, results, keyorder.get)

ValueError: could not convert string to float: S

Traceback
---------
  File "C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\async.py", line 267, in execute_task
    result = _execute_task(task, data)
  File "C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\async.py", line 249, in _execute_task
    return func(*args2)
  File "C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\dataframe\csv.py", line 49, in bytes_read_csv
    coerce_dtypes(df, dtypes)
  File "C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\dataframe\csv.py", line 73, in coerce_dtypes
    df[c] = df[c].astype(dtypes[c])
  File "C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\pandas\core\generic.py", line 2632, in astype
    'argument' % target)
  File "C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\pandas\core\internals.py", line 2864, in astype
    # fillna internally does putmask, maybe it's better to do this
  File "C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\pandas\core\internals.py", line 2823, in apply
    integrity check
  File "C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\pandas\core\internals.py", line 430, in astype
  File "C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\pandas\core\internals.py", line 472, in _astype
  File "C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\pandas\core\common.py", line 2466, in _astype_nansafe
```

**Comment by mrocklin** [comment]
Trying that out locally on that sample everything works fine.  Can you provide a minimal reproducible failing example?

**Comment by kumarivin** [comment]
https://drive.google.com/file/d/0B_XeGxixAToSTnV3R1d0UDJERUU/view?usp=sharing 

Please find attached the minimal data that projects the error

**Comment by mrocklin** [comment]
Hrm, this also works fine on my machine.  I tried it out on a Windows VM and also wasn't able to reproduce your error.

Perhaps you could use the python debugger to figure out what the Pandas error you're getting is about?  To do this I would try your computation in ipython after running the `%pdb` magic.  You will want to set dask to use the single-threaded scheduler by default

    import dask
    dask.set_options(get=dask.async.get_sync)

**Comment by kumarivin** [comment]
Attaching the new error below

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-12-6bcecb57c7b2> in <module>()
      8 dt = df.dtypes.to_dict()
      9 df = dd.read_csv('C:\Users\TG\Downloads\yoochoose-dataFull\yoochoose-clicks.dat', header=None, dtype=dt)
---> 10 get_ipython().magic(u'time len(df)')

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\IPython\core\interactiveshell.pyc in magic(self, arg_s)
   2161         magic_name, _, magic_arg_s = arg_s.partition(' ')
   2162         magic_name = magic_name.lstrip(prefilter.ESC_MAGIC)
-> 2163         return self.run_line_magic(magic_name, magic_arg_s)
   2164 
   2165     #-------------------------------------------------------------------------

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\IPython\core\interactiveshell.pyc in run_line_magic(self, magic_name, line)
   2082                 kwargs['local_ns'] = sys._getframe(stack_depth).f_locals
   2083             with self.builtin_trap:
-> 2084                 result = fn(*args,**kwargs)
   2085             return result
   2086 

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\IPython\core\magics\execution.pyc in time(self, line, cell, local_ns)

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\IPython\core\magic.pyc in <lambda>(f, *a, **k)
    191     # but it's overkill for just that one bit of state.
    192     def magic_deco(arg):
--> 193         call = lambda f, *a, **k: f(*a, **k)
    194 
    195         if callable(arg):

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\IPython\core\magics\execution.pyc in time(self, line, cell, local_ns)
   1171         if mode=='eval':
   1172             st = clock2()
-> 1173             out = eval(code, glob, local_ns)
   1174             end = clock2()
   1175         else:

<timed eval> in <module>()

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\dataframe\core.pyc in __len__(self)
    286 
    287     def __len__(self):
--> 288         return reduction(self, len, np.sum, token='len').compute()
    289 
    290     def map_partitions(self, func, columns=no_default, *args, **kwargs):

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\base.pyc in compute(self, **kwargs)
     84             Extra keywords to forward to the scheduler ``get`` function.
     85         """
---> 86         return compute(self, **kwargs)[0]
     87 
     88     @classmethod

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\base.pyc in compute(*args, **kwargs)
    177         dsk = merge(var.dask for var in variables)
    178     keys = [var._keys() for var in variables]
--> 179     results = get(dsk, keys, **kwargs)
    180 
    181     results_iter = iter(results)

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\async.pyc in get_sync(dsk, keys, **kwargs)
    521     queue = Queue()
    522     return get_async(apply_sync, 1, dsk, keys, queue=queue,
--> 523                      raise_on_exception=True, **kwargs)
    524 
    525 

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\async.pyc in get_async(apply_async, num_workers, dsk, result, cache, queue, get_id, raise_on_exception, rerun_exceptions_locally, callbacks, **kwargs)
    488             f(key, res, dsk, state, worker_id)
    489         while state['ready'] and len(state['running']) < num_workers:
--> 490             fire_task()
    491 
    492     # Final reporting

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\async.pyc in fire_task()
    459         # Submit
    460         apply_async(execute_task, args=[key, dsk[key], data, queue,
--> 461                                         get_id, raise_on_exception])
    462 
    463     # Seed initial tasks into the thread pool

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\async.pyc in apply_sync(func, args, kwds)
    509 def apply_sync(func, args=(), kwds={}):
    510     """ A naive synchronous version of apply_async """
--> 511     return func(*args, **kwds)
    512 
    513 

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\async.pyc in execute_task(key, task, data, queue, get_id, raise_on_exception)
    265     """
    266     try:
--> 267         result = _execute_task(task, data)
    268         id = get_id()
    269         result = key, result, None, id

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\async.pyc in _execute_task(arg, cache, dsk)
    247         func, args = arg[0], arg[1:]
    248         args2 = [_execute_task(a, cache) for a in args]
--> 249         return func(*args2)
    250     elif not ishashable(arg):
    251         return arg

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\dataframe\csv.pyc in bytes_read_csv(b, header, kwargs, dtypes, columns, write_header, enforce)
     47     df = pd.read_csv(bio, **kwargs)
     48     if dtypes:
---> 49         coerce_dtypes(df, dtypes)
     50 
     51     if enforce and columns and (list(df.columns) != list(columns)):

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\dask\dataframe\csv.pyc in coerce_dtypes(df, dtypes)
     71                     raise TypeError("Runtime type mismatch. "
     72                     "Add {'%s': float} to dtype= keyword in read_csv" % c)
---> 73             df[c] = df[c].astype(dtypes[c])
     74 
     75 

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\pandas\core\generic.pyc in astype(self, dtype, copy, raise_on_error, **kwargs)
   2948 
   2949         mgr = self._data.astype(dtype=dtype, copy=copy,
-> 2950                                 raise_on_error=raise_on_error, **kwargs)
   2951         return self._constructor(mgr).__finalize__(self)
   2952 

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\pandas\core\internals.pyc in astype(self, dtype, **kwargs)
   2936 
   2937     def astype(self, dtype, **kwargs):
-> 2938         return self.apply('astype', dtype=dtype, **kwargs)
   2939 
   2940     def convert(self, **kwargs):

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\pandas\core\internals.pyc in apply(self, f, axes, filter, do_integrity_check, consolidate, raw, **kwargs)
   2888 
   2889             kwargs['mgr'] = self
-> 2890             applied = getattr(b, f)(**kwargs)
   2891             result_blocks = _extend_blocks(applied, result_blocks)
   2892 

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\pandas\core\internals.pyc in astype(self, dtype, copy, raise_on_error, values, **kwargs)
    432                **kwargs):
    433         return self._astype(dtype, copy=copy, raise_on_error=raise_on_error,
--> 434                             values=values, **kwargs)
    435 
    436     def _astype(self, dtype, copy=False, raise_on_error=True, values=None,

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\pandas\core\internals.pyc in _astype(self, dtype, copy, raise_on_error, values, klass, mgr, **kwargs)
    475 
    476                 # _astype_nansafe works fine with 1-d only
--> 477                 values = com._astype_nansafe(values.ravel(), dtype, copy=True)
    478                 values = values.reshape(self.shape)
    479 

C:\Users\TG\Anaconda3\envs\dato-env\lib\site-packages\pandas\core\common.pyc in _astype_nansafe(arr, dtype, copy)
   1918 
   1919     if copy:
-> 1920         return arr.astype(dtype)
   1921     return arr.view(dtype)
   1922 

ValueError: could not convert string to float: S

> c:\users\tg\anaconda3\envs\dato-env\lib\site-packages\pandas\core\common.py(1920)_astype_nansafe()
   1918 
   1919     if copy:
-> 1920         return arr.astype(dtype)
   1921     return arr.view(dtype)
   1922 
```

**Comment by mrocklin** [comment]
I'm actually suggesting that you look through some of those variables and see if you can get a clearer picture of what's going on.  We're giving Pandas some values that it doesn't like.  It's not clear to me how this is happening and unfortunately I'm not able to reproduce your error on any of my machines.

**Comment by kumarivin** [comment]
@mrocklin i understood what went wrong, the fourth column in my data is a mixed datatype, once I converted that to np.object type, dask is working fine

**Comment by frol** [comment]
It seems like the issue is already resolved in #1420, so the next Dask release should already include the fix.
